hw09

zuwarka opened the pull request.

zuwarka committed:
Create README.md

zuwarka committed:
Add files via upload

zuwarka committed:
Create html_builder.rb

Предварительная версия html_builder. 
Вывод следующий: 
<html>
  <body>
    <h1 class="h1-formatted margin-10">Hello!</h1>
    <p>This is paragraph<p>Paragraph inside paragraph</p>
</p>
    <div>New div element<p>Paragraph in div</p>
</div>
    <div class="empty-div"/>
  </body>
</html>

zuwarka committed:
Create README.md

zuwarka committed:
Update README.md

zuwarka committed:
Update README.md

zuwarka committed:
Update README.md

Alex6Alex commented on homework10 (12.08.2022)/README.md:
```diff
@@ -0,0 +1,19 @@
+# Задание 1 - html_builder
```
Это задание стоило сделать с новой ветке, можно было в той, где велась работа с обновлением списка

Alex6Alex commented on homework10 (12.08.2022)/html_builder.rb:
```diff
@@ -0,0 +1,79 @@
+class HtmlBuilder
+  def initialize(&block)
+    @result = ""
+    @space = 2
```
Можно было сделать проще, без отступов. Для добавления отступов скорее бы подошел какой-то отдельный класс с описанием форматирования. А обязанность этого класса - просто сделать структуру

Alex6Alex commented on homework10 (12.08.2022)/html_builder.rb:
```diff
@@ -0,0 +1,79 @@
+    @result != "" ? @result = "<html>\n#{@result}</html>" : @result
+  end
+
+  def body(&block)
```
Думаю, что этот метод лишний, так как body такой же элемент, как и p, div с точки зрения тегов, если не рассматривать значение каждого тега

Alex6Alex commented on homework10 (12.08.2022)/html_builder.rb:
```diff
@@ -0,0 +1,79 @@
+  def method_missing(method_name, *args, class_name: nil, &block)
+    @result += "  " * @space + "<#{method_name}"
+
+    unless class_name.nil?
```
`@result += " class=\"#{class_name}\"" unless class_name.nil?`

Alex6Alex replied:
Не уверен, нужны ли тут слешы

Alex6Alex commented on homework10 (12.08.2022)/html_builder.rb:
```diff
@@ -0,0 +1,79 @@
+      @result += " class=\"#{class_name}\""
+    end
+
+    if args.empty? and !block_given?
```
В руби логика работы `and` немного отличается от `&&`, поэтому стоит всегда использовать амперсанды. Рубокоп, например, в большинстве случаев будет ругаться на такую запись. https://mixandgo.com/learn/ruby/boolean-operator-precedence. And чаще всего удобен в последовательности действий, чем в условиях

Alex6Alex commented on homework10 (12.08.2022)/html_builder.rb:
```diff
@@ -0,0 +1,79 @@
+      @space += 2
+    end
+
+    if !args.empty? or block_given?
```
С `or` и `||` та же ситуация

Alex6Alex commented on homework09 (05.08.2022)/circle.rb:
```diff
@@ -0,0 +1,10 @@
+require_relative 'figures'
+
+class Circle < Figure
+  include Math
```
Лучше использовать Math::PI вместо подключения модуля, иначе получается, что интерфейс класса расширился большим количеством математических методов

Alex6Alex commented on homework09 (05.08.2022)/circle.rb:
```diff
@@ -0,0 +1,10 @@
+
+class Circle < Figure
+  include Math
+  attr_writer :circle_r
```
attr_writer не используется ни внутри, ни снаружи (более того, снаружи и не стоит его использовать)

Alex6Alex commented on homework09 (05.08.2022)/figures.rb:
```diff
@@ -0,0 +1,14 @@
+class Figure
+  attr_accessor :area
```
Думаю, достаточно было бы attr_reader'a

Alex6Alex commented on homework09 (05.08.2022)/custom_shape.rb:
```diff
@@ -0,0 +1,7 @@
+require_relative 'figures'
+
+class CustomShape < Figure
+  def initialize(my_area)
```
Этот метод тоже наследуется, поэтому необязательно его дублировать

Alex6Alex commented on homework09 (05.08.2022)/figures.rb:
```diff
@@ -0,0 +1,14 @@
+    @area = my_area
+  end
+
+  def <=>(other_shape)
```
Хорошо бы еще сделать include Comparable, чтобы была возможность полноценно сравнивать эти объекты. А так, идея правильная

Alex6Alex commented on homework09 (05.08.2022)/figures.rb:
```diff
@@ -0,0 +1,14 @@
+    self.area <=> other_shape.area
+  end
+
+  def to_s
```
Этот метод, пожалуй, лишний.

Alex6Alex commented on homework09 (05.08.2022)/rectangle.rb:
```diff
@@ -0,0 +1,10 @@
+require_relative 'figures'
+
+class Rectangle < Figure
+  attr_writer :rectangle_a, :rectangle_b
```
Лишний attr_writer

Alex6Alex commented on homework09 (05.08.2022)/square.rb:
```diff
@@ -0,0 +1,9 @@
+require_relative 'figures'
+
+class Square < Figure
+  attr_writer :square_a
```
И тут

Alex6Alex commented on homework09 (05.08.2022)/triangle.rb:
```diff
@@ -0,0 +1,10 @@
+require_relative 'figures'
+
+class Triangle < Figure
+  attr_writer :triangle_a, :triangle_h
```
И тут

zuwarka committed:
Update html_builder.rb

Отредактирован класс HtmlBuilder согласно правкам: убран отступ, рефакторинг проверки условия на nil class_name, заменены and и or на && и || соответственно.

zuwarka committed:
Update circle.rb

Убран include Math и attr_

zuwarka committed:
Update custom_shape.rb

Убрано лишнее переопределение конструктора

zuwarka committed:
Update figures.rb

Убран метод to_s, добавлена библиотека Comparable

zuwarka committed:
Update rectangle.rb

Убран attr_

zuwarka committed:
Update square.rb

Убран attr_

zuwarka committed:
Update triangle.rb

Убран attr_
